cmd/clusterctl/internal/migrate/engine.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migrate
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// Engine orchestrates the migration process by combining a parser and converter.
+type Engine struct {
+	parser    *YAMLParser
+	converter *Converter
+
+	errs     []error
+	warnings []string
+}
+
+// MigrationOptions contains configuration for a migration run.
+type MigrationOptions struct {
+	Input     io.Reader
+	Output    io.Writer
+	Errors    io.Writer
```

Comment:
Considering we have to model this for consumption as library I don't think we should use io.Reader.
Similarly, let's not use io.Writer as for output and errors (and move them to the output of the migration function)

```suggestion
	Input     []byte
```